Mathlib/Probability/ProductMeasure.lean

```diff
@@ -0,0 +1,463 @@
+  intro I
+  ext s hs
+  rw [map_apply (measurable_restrict I) hs, infinitePi, AddContent.measure_eq, ← cylinder,
+    piContent_cylinder μ hs]
+  · exact generateFrom_measurableCylinders.symm
+  · exact cylinder_mem_measurableCylinders _ _ hs
+
+theorem infinitePi_map_restrict {I : Finset ι} :
+    (Measure.infinitePi μ).map I.restrict = Measure.pi fun i : I ↦ μ i :=
+  isProjectiveLimit_infinitePi μ I
+
+instance : IsProbabilityMeasure (infinitePi μ) := by
+  constructor
+  rw [← cylinder_univ ∅, cylinder, ← map_apply (measurable_restrict _) .univ,
+    infinitePi_map_restrict, measure_univ]
+
+theorem eq_infinitePi {ν : Measure (Π i, X i)}
+    (hν : ∀ s : Finset ι, ∀ t : (i : ι) → Set (X i),
+      (∀ i ∈ s, MeasurableSet (t i)) → ν (Set.pi s t) = ∏ i ∈ s, μ i (t i)) :
+    ν = infinitePi μ := by
+  refine (isProjectiveLimit_infinitePi μ).unique ?_ |>.symm
+  refine fun s ↦ (pi_eq fun t ht ↦ ?_).symm
+  classical
+  rw [Measure.map_apply, restrict_preimage, hν, ← prod_attach, univ_eq_attach]
+  · congr with i
+    rw [dif_pos i.2]
+  any_goals fun_prop
+  · exact fun i hi ↦ dif_pos hi ▸ ht ⟨i, hi⟩
+  · exact .univ_pi ht
+
+-- TODO: add a version for an infinite product
+lemma infinitePi_boxes {s : Finset ι} {t : (i : ι) → Set (X i)}
+    (mt : ∀ i ∈ s, MeasurableSet (t i)) :
+    infinitePi μ (Set.pi s t) = ∏ i ∈ s, (μ i) (t i) := by
+  have : Set.pi s t = cylinder s ((@Set.univ s).pi (fun i : s ↦ t i)) := by
+    ext x
+    simp
+  rw [this, cylinder, ← map_apply, infinitePi_map_restrict, pi_pi]
+  · rw [univ_eq_attach, prod_attach _ (fun i ↦ (μ i) (t i))]
+  · exact measurable_restrict _
+  · exact .univ_pi fun i ↦ mt i.1 i.2
+
+theorem infinitePi_eq_pi [Fintype ι] : infinitePi μ = Measure.pi μ := by
+  refine (pi_eq fun s hs ↦ ?_).symm
+  rw [← coe_univ, infinitePi_boxes]
+  simpa
+
+lemma infinitePi_cylinder {s : Finset ι} {S : Set (Π i : s, X i)} (mS : MeasurableSet S) :
+    infinitePi μ (cylinder s S) = Measure.pi (fun i : s ↦ μ i) S := by
+  rw [cylinder, ← Measure.map_apply (measurable_restrict _) mS, infinitePi_map_restrict]
+
+end Measure
+
+section Integral
+
+theorem integral_restrict_infinitePi {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+    {s : Finset ι} {f : (Π i : s, X i) → E}
+    (hf : AEStronglyMeasurable f (Measure.pi (fun i : s ↦ μ i))) :
+    ∫ y, f (s.restrict y) ∂infinitePi μ = ∫ y, f y ∂Measure.pi (fun i : s ↦ μ i) := by
+  rw [← integral_map, infinitePi_map_restrict]
+  · fun_prop
+  · rwa [infinitePi_map_restrict]
+
+theorem lintegral_restrict_infinitePi {s : Finset ι}
+    {f : (Π i : s, X i) → ℝ≥0∞} (hf : Measurable f) :
+    ∫⁻ y, f (s.restrict y) ∂infinitePi μ = ∫⁻ y, f y ∂Measure.pi (fun i : s ↦ μ i) := by
+  rw [← lintegral_map hf (measurable_restrict _), isProjectiveLimit_infinitePi μ]
+
+open Filtration
+
+theorem StronglyMeasurable.integral_of_piFinset [DecidableEq ι] {E : Type*} [NormedAddCommGroup E]
```

Comment:
I don't think I'd put `StronglyMeasurable` as the main namespace here, but on the other hand I'd mention `infinitePi` somewhere in the name. Same thing in the next theorem.